gufe/tests/test_custom_json.py

```diff
@@ -0,0 +1,145 @@
+# This file was originally part of OpenPathSampling/SimStore, which is
+# distributed under the MIT license.
+# Portions Copyright (c) 2014-2022 the contributors to OpenPathSampling
+# Permissions are the same as those listed in the gufe LICENSE
+
+from gufe.custom_json import (
+    JSONSerializerDeserializer, custom_json_factory, JSONCodec, PATH_CODEC
+)
+import json
+import pathlib
+import pytest
+
+import numpy as np
+from numpy import testing as npt
+
+bytes_codec = JSONCodec(
+    cls=bytes,
+    to_dict=lambda obj: {'latin-1': obj.decode('latin-1')},
+    from_dict=lambda dct: dct['latin-1'].encode('latin-1'),
+)
+
+numpy_codec = JSONCodec(
+    cls=np.ndarray,
+    to_dict=lambda obj: {
+        'dtype': str(obj.dtype),
+        'shape': list(obj.shape),
+        'bytes': obj.tobytes()
+    },
+    from_dict=lambda dct: np.frombuffer(
+        dct['bytes'], dtype=np.dtype(dct['dtype'])
+    ).reshape(dct['shape'])
+)
```

Comment:
This is a good example, but we also need these built in as "batteries included" components, along with openff.units.  Can we merge these into the actual library and keep the existing tests (that require them built in), then separately have a dumb class we define in this test file to show the extension of Serializer.

Reply:
& Pathlib.path

Reply:
Sure, I'll move them over. I didn't know if we wanted to handle numpy objects; so I only put them in tests.

Do we really want a bytes codec? Or should I make it so that the numpy codec just returns string directly?

It's kind of a question of how much non-JSON standard do we want to enable storage of: tuples? sets? dicts with non-string keys? All of this is possible; just a questions of what is worth the time to support (OPS supports tuples and sets, so implementation exists if we want them).